### ISSUE TITLE
[Due for payment 2025-03-21] [$500] Feature request : Upgrade task titles with markdown rendering

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#53175]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:**
**Reproducible in staging?:** New Feature
**Reproducible in production?:** New Feature
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:**
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @quinthar 
**Slack conversation** (hyperlinked to channel name): [ts_external_expensify_convert](https://expensify.slack.com/archives/C07HPDRELLD/p1732652969906309?thread_ts=1732105057.984919&cid=C07HPDRELLD)

**Background:** The primary use of tasks today is Concierge assigning a specific set of onboarding task to the user.  Each task currently has a numbered list of steps, generally ending with a link we want them to click.  However, we see that some users don't realize the task can be clicked, meaning they don't see the instructions, meaning they don't click the link.

**Problem:** When new users don't realize the task title is clickable, they don't see the instructions, and thus don't do the call to action.

**Solution:** Please render task names using the same markdown>html processing as we do normal comments, to allow hyperlinks to be put into titles, in order to shorten the path to the CTA to one click.  In other words, make task titles work like task descriptions:

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Standalone
- [ ] Android: HybridApp
- [ ] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

<details>
  <summary>Add any screenshot/video evidence</summary>

  
</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021863958075863531330</li>
        <li>Upwork Job ID: 1863958075863531330</li>
        <li>Last Price Increase: 2024-12-07</li>
<li>Automatic offers: </li>
<ul>
<li>Krishna2323 | Contributor | 105277373</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @JmillsExpensify</details>

**Comment by quinthar** [comment]
Hm, why hasn't this had any proposals?  Also, why is the job listing "private"?

**Comment by mallenexpensify** [comment]
The job shows as [public to me in Upwork](https://www.upwork.com/jobs/~021863958075863531330). 
Likely no proposals cuz it's at the lowest price, I bumped to $500 to get 👀
If no proposals are submitted by Monday I can try to rally a contributor.

**Comment by mallenexpensify** [comment]
@mananjadhav 👀 plz on the proposal above.  Thx

**Comment by mananjadhav** [comment]
Yes I had reviewed the proposal partially on the weekend. The proposal looks good, minor code changes and other areas to make updates can be covered in the implementation. 

🎀 👀 🎀

**Comment by quinthar** [comment]
> @johnmlee101 @mananjadhav, where should we store the markdown text (task title)? The task title is currently stored in the reportName property of the report, but this property does not support markdown syntax and has a maximum length of 100 characters. This length will not be sufficient for the task title, as it will include links in markdown syntax as well.

Why doesn't the task title support markdown; the purpose of this PR is to make it support markdown.  

Also, why does it have a maximum length of 100 characters -- where is that enforced, in the API?

**Comment by johnmlee101** [comment]
https://github.com/Expensify/Auth/blob/afe8b9deefd8de62a088f5349d64801ac6303ca3/auth/lib/Request.h#L27 we have a limit of 10k in CreateReport so not sure where this limit is. What error are you getting if you set it to something greater?

**Comment by johnmlee101** [comment]
Ah okay, I find it now. This is the title length of all rooms so I think making it larger might introduce a lot of unintended side effects

**Comment by johnmlee101** [comment]
So my next recommendation is potentially storing these markdown titles in the report NVPs themselves. We can have a `markdownTitle` potentially? I would need to see the lifecycle of loading reports to make sure that itself doesn't load independently to the title today

**Comment by quinthar** [comment]
I don't think I understand that.  We are rendering it where the title is; when this is set, that's the title, there is no other title.  I don't think I understand why we can't/won't store it in the title.

**Comment by quinthar** [comment]
One thing that occurs to me is that we kind of render task titles in two ways:

1) Everywhere else a report title appears, such as in the header.  In this case, it's already "clickable", because it opens the task details:

<img width="1440" alt="image" src="https://github.com/user-attachments/assets/b87827e9-9275-402e-88f6-9a473c6db1fd" />

2) In a "task preview", such as when it's in a chat:

<img width="1440" alt="image" src="https://github.com/user-attachments/assets/80b7a442-a970-41b6-b734-3a8370bacf1c" />


I think we actually only want to show the hyperlinked URL in (2).  Given that, how about we:

* Store the HTML version in a reportNameValuePair
* Store a text-only version in the reportTitle

This avoids us needing to refactor every single place a task title can appear, and instead *only* display the HTML version in the "task preview".  Thoughts on this simplification?

**Comment by quinthar** [comment]
Raised realtime discussion here: https://expensify.slack.com/archives/C01GTK53T8Q/p1734840786675329

**Comment by quinthar** [comment]
Put updated plan here: https://expensify.slack.com/archives/C01GTK53T8Q/p1734899964964939?thread_ts=1734840786.675329&cid=C01GTK53T8Q

**Comment by mallenexpensify** [comment]
@quinthar asked @johnmlee101 to confirm the plan in [thread here](https://expensify.slack.com/archives/C01GTK53T8Q/p1735352281932559?thread_ts=1734840786.675329&cid=C01GTK53T8Q).

**Comment by quinthar** [comment]
Bumped @johnmlee101 for an update.

**Comment by johnmlee101** [comment]
WIP PR ongoing

**Comment by mananjadhav** [comment]
Can we put this on Hold till we get the BE PR merged?

**Comment by mallenexpensify** [comment]
Updated preface to title
> [HOLD BE Fix]

**Comment by quinthar** [comment]
@johnmlee101 What's your ETA?

**Comment by johnmlee101** [comment]
@Krishna2323 I'm currently using your solution to partially allow for HTML report titles. I'll then open it up so that your PR will allow for rendering the HTML properly.

**Comment by johnmlee101** [comment]
Both PRs are up in review right now

**Comment by johnmlee101** [comment]
Currently waiting for web PR so we can test this on staging

**Comment by Shahidullah-Muffakir** [comment]
Hi everyone,
Since we're already localizing the Tasks https://github.com/Expensify/App/issues/56361 (title, description, and message), do we still need to implement these changes?

It seems we can simply display the translated markdown directly from the `en.ts `and `es.ts `files, based on the translation keys used when creating the tasks after onboarding

**Comment by johnmlee101** [comment]
We will still need to support markdown to better support more customized tasks

**Comment by roryabraham** [comment]
Sadly, @johnmlee101 is no longer working at Expensify, but the back-end part of this is done so I'm going to open it up to external volunteers to take the front-end part over the finish line. John got pretty far in https://github.com/Expensify/App/pull/55967, so that's a good starting point.

**Comment by mananjadhav** [comment]
Thanks for the update @roryabraham.

**Comment by mananjadhav** [comment]
Thanks for the update @Krishna2323

**Comment by shawnborton** [comment]
For the title shown here:

![Image](https://github.com/user-attachments/assets/ecee4946-ed0d-4d47-a21d-1e1e075da7f3)

Can we make sure that we are honoring the existing headline style that is being used for task titles? This means that we should keep the title looking like this, even if markdown was applied:

![Image](https://github.com/user-attachments/assets/8f399b4a-4961-47a0-8a93-ee31cb912431)

Mostly I just want to make sure we aren't modifying the font size or font weight here, since we technically already have a headline style happening. Thanks! And cc @Expensify/design for viz

**Comment by shawnborton** [comment]
> @Expensify/design, does this look better? ![Image](https://github.com/user-attachments/assets/bb12475b-bb6b-433b-94ac-6a07a2b91e93)

Hmm, part of me thinks italic text should just use our italic headline style (as in, the font should be Expensify New Kansas but in italics). For the bold text... I am stumped there. We only use 500 weight for Expensify New Kansas, which is medium, but we could potentially supply a heavier weight of ENK for these specific moments. Thoughts @Expensify/design? Or we could be lazy and just let the browser add the boldness, which I don't love but it allows us to not have to supply more font files.

**Comment by dubielzyk-expensify** [comment]
Hmm this is tricky. I feel like we allow markdown in the titles, then we should abide by it, but also that means `code` too right? It just gets super messy. I feel like we should either do bold, italic, line through, etc, or do links only. It just feels a bit weird if it's a guessing game. I highly doubt this'll be used outside of our own use cases for this anyways.

**Comment by shawnborton** [comment]
> but also that means code too right?

Agree. GitHub's issue titles allow `inline code` but no other styling AFAIK. We could take a similar approach, but also allow links since that's more relevant in our use case.

**Comment by dubielzyk-expensify** [comment]
I don't mind that approach personally

**Comment by dannymcclain** [comment]
I think I prefer that approach personally. [This image](https://github.com/Expensify/App/issues/53175#issuecomment-2669876620) looks so messy and TBH I don't really see the value in allowing full-on markdown in a task **title**. For the description, which is more likely to be long form-ish, sure, let's go wild. But for a title, I see no reason not to clamp down on the allowed styles (like many other apps do). `:my-2-cents:`

**Comment by shawnborton** [comment]
Cool, I say let's do it then. So the task titles, when viewing the task it self, would only show styles for links or `inline code` but not bold or italic. I think it's fine to support strikethrough since it doesn't require a new font but I don't feel strongly.

**Comment by dannymcclain** [comment]
Agree!

**Comment by dubielzyk-expensify** [comment]
Sounds good to me. Think it's also nicer to not have a new font added due to extra app size.

**Comment by shawnborton** [comment]
> do we also need to disable the markdown styles in edit mode? Thanks!

I wouldn't change or disable anything here. 

For this screenshot here:

![Image](https://github.com/user-attachments/assets/ea3f7acd-f80a-4050-b940-2a75a9d87d49)

I would think that you shouldn't see the raw characters that enable markdown (like the hashtag, the asterisk, etc) - the only thing we're requesting here is that we just don't render those particular styles. cc @Expensify/design for the gut check.

**Comment by shawnborton** [comment]
Also, FWIW, I think we do have italic styles for our headline font, so I would be cool including italics in the title.

**Comment by dannymcclain** [comment]
>do we also need to disable the markdown styles in edit mode? Thanks!

>>I wouldn't change or disable anything here.

I actually do think we should not preview the markdown styles that won't show up in the title. I think it's confusing to type out bold and strike through and have those styles show up in the edit field but not in the title.

>I would think that you shouldn't see the raw characters that enable markdown (like the hashtag, the asterisk, etc) - the only thing we're requesting here is that we just don't render those particular styles

I don't think we would want to _remove_ any characters that the user typed—if they type hashtags and asterisks, they wouldn't _do_ anything, but I don't see why we'd remove them.

**Comment by shawnborton** [comment]
> I think it's confusing to type out bold and strike through and have those styles show up in the edit field but not in the title.

But those styles would show up from the chat view... I also don't love the idea of having certain text inputs that have different rules for how markdown is entered. 

> I don't think we would want to remove any characters that the user typed—if they type hashtags and asterisks, they wouldn't do anything, but I don't see why we'd remove them.

Interesting, I guess this gets tricky because those styles would show up from the chat view, right? Like if you go to the parent chat with the task.

**Comment by dannymcclain** [comment]
>But those styles would show up from the chat view... I also don't love the idea of having certain text inputs that have different rules for how markdown is entered.

Ah dip I didn't think about this. You're right.

>Interesting, I guess this gets tricky because those styles would show up from the chat view, right? Like if you go to the parent chat with the task.

Again, yes. I forgot about the chat view haha. Probably best to just do the least amount of messing about with this then, yeah?

**Comment by shawnborton** [comment]
That's kind of what I'm thinking. Basically we just don't render bold styles or heading styles in the task title when you view the task itself, but we probably can support all the other cases for markdown there.

**Comment by dannymcclain** [comment]
Ok yeah that sounds good to me. So basically, ignore everything I said above and just listen to Shawn 😂

**Comment by dubielzyk-expensify** [comment]
I was also very much thinking along the same lines as you, Danny, but I think what Shawn is saying is very reasonable and makes sense 😄

**Comment by shawnborton** [comment]
@Krishna2323 is there a way to basically allow all markdown to be entered in the edit input, and just not show any new styles when we detect bold text in a task's title when you view the task itself? All of the other styles should check out just fine I think.

**Comment by shawnborton** [comment]
> Do we need to show the bold/heading style in preview?

Yeah, keep everything the same regarding how the markdown is added to the editable text input. The only thing we won't do is make any custom styles for `#` or `*` when you view the title from the task itself. Let me know if that makes sense!

**Comment by shawnborton** [comment]
What do you mean by that exactly?

**Comment by shawnborton** [comment]
Yup, I think we should use all markdown available to us there. The same way we would add it to any other chat. 

My main thinking behind not doing anything special for the bold and heading styles when you view the task itself is that in that particular view, the task title already has a special "heading" style, which uses our special flavor of bold (aka medium).

**Comment by shawnborton** [comment]
Sounds great, keep us posted and we'll help out with the design reviews!

**Comment by dannymcclain** [comment]
Nice! I think that's exactly what @shawnborton was describing.

**Comment by shawnborton** [comment]
Yes, that looks great to me! Can you show us what some `inline code` looks like as well just for posterity?

**Comment by shawnborton** [comment]
Nice, that seems good to me - thanks!

**Comment by mananjadhav** [comment]
PR is WIP.

**Comment by quinthar** [comment]
Looking for engineering volunteer: https://expensify.slack.com/archives/C03TQ48KC/p1740777106523109

**Comment by dubielzyk-expensify** [comment]
1. I don't think we should change anything from what it is today, so I like keeping it strike-through on completed tasks like it is today. It does become a bit weird with the code blocks so I could see the argument for removing it and if we did we could either A) keep it same text styling with and without or B) we could apply text-supporting to it to make it more fade out a bit more.
2. I think the sizing of the code block example above look pretty good. I guess we should also support multi-line code as well given we support everything else. 

Keen on @Expensify/design thoughts as well. This is becoming pretty hectic, but I guess it's still working. Part of me still kinda wishes we just stuck with links and italics 😂

**Comment by shawnborton** [comment]
Yup, I agree with everything Jon is saying above. And yeah, I agree that this is feeling slightly silly... but maybe we start with allowing full markdown control here, and then we can scale back if we find it to be an issue. I have a feeling most of our users don't even use markdown or aren't familiar with it.

**Comment by dubielzyk-expensify** [comment]
Yeah, I'm okay with that 👍

**Comment by dannymcclain** [comment]
I also agree with Jon and Shawn's thoughts. I'm fine starting with all of it and then scaling back.

**Comment by quinthar** [comment]
Hi!  This project seems stuck, so I've asked @Krishna2323 some questions in Slack here: https://expensify.slack.com/archives/C01GTK53T8Q/p1741054265831269?thread_ts=1734840786.675329&cid=C01GTK53T8Q  -- I'd love to get a better understanding of where we are currently at with the front end, so we can identify exactly what needs to happen on the back end.

**Comment by quinthar** [comment]
Seems like @roryabraham has reviewed https://github.com/Expensify/App/pull/54165 and left comments, and @Krishna2323 is responding to those comments.  Once resolved, we can merge.

**Comment by quinthar** [comment]
Bumped again -- I think we should ship what we've got, remaining issues aren't blockers.

**Comment by quinthar** [comment]
Merged and going out on Monday!

**Comment by VictoriaExpensify** [comment]
# BugZero Checklist:

- [ ] **[Contributor]** Classify the bug:

<details>
<summary>Bug classification</summary>


Source of bug:
  - [ ] 1a. Result of the original design (eg. a case wasn't considered)
  - [ ] 1b. Mistake during implementation
  - [ ] 1c. Backend bug
  - [ ] 1z. Other:

Where bug was reported:
  - [ ] 2a. Reported on production (eg. bug slipped through the normal regression and PR testing process on staging)
  - [ ] 2b. Reported on staging (eg. found during regression or PR testing)
  - [ ] 2d. Reported on a PR
  - [ ] 2z. Other:

Who reported the bug:
  - [ ] 3a. Expensify user
  - [ ] 3b. Expensify employee
  - [ ] 3c. Contributor
  - [ ] 3d. QA
  - [ ] 3z. Other:

</details>

- [ ] **[Contributor]** The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake.

    Link to comment:

- [ ] **[Contributor]** If the regression was CRITICAL (e.g. interrupts a core flow) A discussion in [#expensify-open-source](https://app.slack.com/client/E047TPA624F/C01GTK53T8Q) has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner.

    Link to discussion:

- [ ] **[Contributor]** If it was decided to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps using the template below to ensure the same bug will not reach production again.

<details>
<summary>Regression Test Proposal Template</summary>


- [ ] **[BugZero Assignee]** Create a GH issue for creating/updating the regression test once above steps have been agreed upon.

    Link to issue:

## Regression Test Proposal
### Precondition:


-

### Test:


1.

Do we agree 👍 or 👎


</details>

**Comment by VictoriaExpensify** [comment]
This wasn't a bug, so the checklist isn't relevant here.

**Comment by VictoriaExpensify** [comment]
Payment summary:
C: @Krishna2323 - paid $500 via Upwork
C+: @mananjadhav - owed $500 via NewDot

**Comment by CortneyOfstad** [comment]
Coming from comments [here](https://github.com/Expensify/App/issues/58192#issuecomment-2766768812) and [here](https://github.com/Expensify/App/issues/53175#issuecomment-2766761484) — reopening

**Comment by quinthar** [comment]
Demoting this to weekly, as we the urgent part is done, and we're in cleanup.

**Comment by mallenexpensify** [comment]
@JmillsExpensify Vic is out, can you please wrap this issue up? Thx

**Comment by mananjadhav** [comment]
@JmillsExpensify Quick bump on this one.

> [@roryabraham](https://github.com/roryabraham) All of the linked PRs are now merged. I was wondering if this could be paid out. Also considering the scope of the issue was much larger than anticipated can we please bump the price for this one?

Also @roryabraham if you got a chance to go through this comment.

**Comment by roryabraham** [comment]
> All of the linked PRs are now merged. I was wondering if this could be paid out. Also considering the scope of the issue was much larger than anticipated can we please bump the price for this one?

This appears ready for payout. I think it would be reasonable to increase the bounty for this one. What price do you think is fair?

**Comment by JmillsExpensify** [comment]
Waiting for final payment amount.

**Comment by roryabraham** [comment]
$1500 seems reasonable to me as well 👍🏼

**Comment by mananjadhav** [comment]
I agree with @Krishna2323's assessment. @JmillsExpensify can you help with the same?

**Comment by mananjadhav** [comment]
Thanks for the update @Krishna2323. I think it's @JmillsExpensify who's going to take care of the payout here.

**Comment by mallenexpensify** [comment]
Contributor: @Krishna2323 paid $1000 via Upwork (already[ paid $500 here]. Payment was on same job, $1k was added as a bonus (https://github.com/Expensify/App/issues/53175#issuecomment-2742354413))
Contributor+: @mananjadhav due $1500 via NewDot. 

Do we need regression tests here? Seems like we would. cc @mananjadhav

**Comment by mallenexpensify** [comment]
> do we have any set of regression tests for markdown fields?

These are the tests I see when I search `Markdown`

<img width="939" alt="Image" src="https://github.com/user-attachments/assets/a9c0367c-df4d-465c-8c94-a0fae2b77b2e" />


Can you access TestRail?  Looks like you have read-only access @mananjadhav 

do we have any set of regression tests for markdown fields?

**Comment by mananjadhav** [comment]
>  Looks like you have read-only access 

Ohh I forgot. I'll have to check. I'll check this on the weekend and try to build the test cases.

**Comment by garrettmknight** [comment]
$1500 approved for @mananjadhav

**Comment by mananjadhav** [comment]
I am copying the tests from individual PRs. Should have them ready by today/tomorrow.

**Comment by mananjadhav** [comment]
### ✅ Regression Test Proposal

1. **Basic Markdown Rendering in Task Title**
   - Go to any chat.
   - Click/tap on `+` > `Assign task`.
   - Paste the following in the title and description fields:
     ```
     This is a sentence without any markdown styling.
     Simple sentence with a [link](https://github.com/Expensify/App/issues) , `inline code` and ~line-through~ text.
     # heading with a [link](https://github.com/Expensify/App/issues) and `inline code`.
     > Blockquote with a [link](https://github.com/Expensify/App/issues) and `inline code`.
     A simple text with a ~line-through~ and _italic_ word.
     # A heading with a line-through and italic word.
     *This sentence is bolded using "**"*
     > We should also check emojis 😅
     # And emojis in a heading 😅
     > Whats missing🤔? Please test the missing ones🥶.
     ```
   - Click "Next" and confirm:
     - Markdown is correctly rendered in title input and matches description.
   - Click "Confirm Task".
   - Validate:
     - Task preview renders markdown properly.
     - Task detail page displays rendered markdown correctly.
     - Text after `#` and between `**` is **not** incorrectly bolded.

2. **Copy Task Title and Validate Markdown Copy Behavior**
   - Create a task with markdown in the title.
   - Right click on the task preview > Copy to clipboard.
   - Paste the content.
   - Ensure the pasted title retains proper markdown formatting.

3. **Blockquote Hover and Rendering Validation**
   - Create a task with blockquote (`> block text`) in the title.
   - Hover on the task preview in chat.
   - Validate:
     - Blockquote line is visible on hover.
     - Clicking task shows the blockquote rendered as normal text in the task view.

4. **Multiline Markdown and Edge Case Formatting**
   - Use multiline markdown in title: `> > hey`, then update to just `hey`.
   - Confirm:
     - Title renders as normal text.
     - No layout issues or excess spacing in task report header.

5. **Image Markdown Handling**
   - Create a task with an inline image markdown in title:
     ```
     [![300](https://picsum.photos/id/1067/200/300)](https://example.com)
     ```
   - Validate:
     - Confirmation page does **not** show image.
     - Task preview shows link instead of image.
     - Task detail view also renders link (no image shown).

6. **Search Router Single-line Enforcement**
   - Create a task with a multiline title.
   - Open the search router (top right).
   - Confirm:
     - Title is rendered as a single line.

7. **Code Block and Link Overlap**
   - Create task with title:
     ```
     [long link name](https://github.com/Expensify/App/issues/google.com)
     inline code
     ```
   - Verify:
     - Inline code does not overlap with the link visually.

8. **Onboarding Flow Consistency**
   - Go through onboarding:
     - Choose “Manage my team’s expenses”
     - Select “1–10 employees”
     - Choose “None”
   - Login using magic link after onboarding.
   - Log in on Android app with same account.
   - Open Concierge chat.
   - Tap any task and verify:
     - Task name remains consistent across navigation.

9. **Strikethrough Style for Completed Tasks**
   - Open a report and create a task.
   - Complete the task.
   - Confirm:
     - Strikethrough is applied in preview for completed task.


@mallenexpensify Can you please add this to the Task title regression test suite?


Once added we're good to close this one.

**Comment by JmillsExpensify** [comment]
Regression tests created so closing this one out.
